packages/core/echo/echo-schema/src/automerge/automerge-db.ts

```diff
@@ -218,36 +231,125 @@ export class AutomergeDb {
     }
   }
 
+  private _initDocAccess(handle: DocHandle<SpaceDoc>) {
+    handle.change((newDoc: SpaceDoc) => {
+      newDoc.access = { spaceKey: this.spaceKey.toHex() };
+    });
+  }
+
+  private _linkObjectDocument(object: AutomergeObject, handle: DocHandle<SpaceDoc>) {
+    this._objectDocumentHandles.set(object.id, handle);
+    this._spaceRootDocHandle.change((newDoc: SpaceDoc) => {
+      if (newDoc.links) {
+        newDoc.links[object.id] = handle.url;
+      } else {
+        newDoc.links = { [object.id]: handle.url };
+      }
+    });
+  }
+
+  private _loadLinkedObjects(links: SpaceDoc['links']) {
+    if (!links) {
+      return;
+    }
+    for (const [objectId, automergeUrl] of Object.entries(links)) {
+      if (this._objectDocumentHandles.has(objectId)) {
+        return;
+      }
+      const handle = this.automerge.repo.find<SpaceDoc>(automergeUrl as DocumentId);
+      log.debug('document loading triggered', { objectId, automergeUrl });
+      this._objectDocumentHandles.set(objectId, handle);
+      this._createObjectOnDocumentLoad(objectId, handle);
+    }
+  }
+
+  private _onDocumentUpdate(event: DocHandleChangePayload<SpaceDoc>) {
+    const { updatedObjects, linkedDocuments } = processDocumentUpdate(event);
+    this._loadLinkedObjects(linkedDocuments);
+    this._createInlineObjects(
+      event.handle,
+      updatedObjects.filter((id) => !this._objects.has(id)),
+    );
+    this._emitUpdateEvent(updatedObjects);
+  }
+
+  private _onDispose() {
+    this._spaceRootDocHandle?.off('change');
+    for (const docHandle of this._objectDocumentHandles.values()) {
+      docHandle.off('change');
```

Comment:
doesn't this remove all listeners? This sounds unsound since AutomergeRepo maintains a map of docHandles to be shared with all consumers